src/core/Akka.Streams/Dsl/DelayFlow.cs

```diff
@@ -0,0 +1,179 @@
+        /// <param name="needsIncrease">if <code>true</code> delay increases, if <code>false</code> delay resets to <paramref name="initialDelay"/></param>
+        /// <param name="initialDelay">initial delay for each of elements</param>
+        /// <param name="maxDelay">limits maximum delay</param>
+        [ApiMayChange]
+        public LinearIncreasingDelay(TimeSpan increaseStep, Func<T, bool> needsIncrease, TimeSpan initialDelay, TimeSpan maxDelay)
+        {
+            if (increaseStep <= TimeSpan.Zero)
+                throw new ArgumentException("Increase step must be positive", nameof(increaseStep));
+
+            if (maxDelay <= initialDelay)
+                throw new ArgumentException("Max delay must be bigger than initial delay");
+
+            _increaseStep = increaseStep;
+            _needsIncrease = needsIncrease;
+            _initialDelay = initialDelay;
+            _maxDelay = maxDelay;
+
+            _delay = _initialDelay;
+        }
+
+        public TimeSpan NextDelay(T elem)
+        {
+            if (_needsIncrease(elem))
+            {
+                var next = _delay + _increaseStep;
+                if (next < _maxDelay)
+                    _delay = next;
+                else
+                    _delay = _maxDelay;
+            }
+            else
+                _delay = _initialDelay;
+
+            return _delay;
+        }
+    }
+
+    /// <summary>
+    /// Flow stage for universal delay management, allows to manage delay through <see cref="IDelayStrategy{T}"/>.
+    /// </summary>
+    /// <typeparam name="T">type of element</typeparam>
+    public class DelayFlow<T> : SimpleLinearGraphStage<T>
+    {
+        #region Logic
+
+        private sealed class Logic : TimerGraphStageLogic
+        {
+            private readonly DelayFlow<T> _delayFlow;
+            private readonly object _delayTimerKey = new object();
+            private T _delayedElem;
+
+            public Logic(DelayFlow<T> delayFlow) : base(delayFlow.Shape)
+            {
+                _delayFlow = delayFlow;
+                var strategy = delayFlow._strategySupplier();
+
+                SetHandler(delayFlow.Inlet, onPush: () =>
+                {
+                    var elem = Grab(delayFlow.Inlet);
+                    var delay = strategy.NextDelay(elem);
+                    if (delay == TimeSpan.Zero)
+                        Push(delayFlow.Outlet, elem);
+                    else
+                    {
+                        _delayedElem = elem;
+                        ScheduleOnce(_delayTimerKey, delay);
+                    }
+                }, onUpstreamFinish: () =>
+                {
+                    if (!IsTimerActive(_delayTimerKey))
+                        CompleteStage();
+                });
+
+                SetHandler(delayFlow.Outlet, onPull: () =>
+                {
+                    Pull(delayFlow.Inlet);
+                });
+            }
```

Comment:
For performance reasons we've introduced `IInHandler` and `IOutHandler` in the past and by implementing those interfaces or one of it's subclasses you can refactor `SetHandler`  to `SetHandler(delayFlow.Inlet, this)` which saves the allocation for the `LamdaHandler` which is created behind the scene.  You can find one example here https://github.com/akkadotnet/akka.net/blob/3f924fda8d6ca55c5af90a710d23be8ce36d5c7a/src/core/Akka.Streams/Implementation/Timers.cs#L53  including the correct implementation for the default calls.